src/ferui-components/tree-view/tree-view-component.ts

```diff
@@ -0,0 +1,425 @@
+      <div *ngIf="error" class="error">
+        <clr-icon class="fui-error-icon" shape="fui-error" aria-hidden="true"></clr-icon>
+        <p>Couldn't load content</p>
+      </div>
+    </div>
+  `,
+  changeDetection: ChangeDetectionStrategy.OnPush,
+  host: {
+    class: 'fui-tree-view-component',
+  },
+})
+export class FuiTreeViewComponent<T> implements OnInit, OnDestroy {
+  @Input() treeNodeData: TreeNodeData<T>;
+
+  @Input() dataRetriever: TreeNodeDataRetriever<T> | PagedTreeNodeDataRetriever<T>;
+
+  @Input() config: TreeViewConfiguration;
+
+  @Input() loading: boolean = false;
+
+  @Input() error: boolean = false;
+
+  @HostBinding('class.show-border') border: boolean;
+
+  @ViewChild(FuiVirtualScrollerComponent) vs: FuiVirtualScrollerComponent;
+
+  @ViewChildren(FuiTreeNodeComponent) children: QueryList<FuiTreeNodeComponent<T>>;
+
+  @Output() onNodeEvent: EventEmitter<TreeViewEvent<T>> = new EventEmitter<TreeViewEvent<T>>();
+
+  treeViewStyles: FuiTreeViewComponentStyles;
+  colorTheme: TreeViewColorTheme;
+  scrollViewArray: TreeNode<T>[];
+  hasBorders: boolean;
+  domBufferAmount: number = 10;
+  private rootNode: TreeNode<T>;
+  private nonRootArray: TreeNode<T>[];
+  private scrollSubscription: Subscription;
+  private scrollWidthChangeSub: Subscription;
+  private originalWidth: number;
+  private scrollPromise: boolean = false;
+  private bufferAmount: number = 1000;
+  private serverSideComponent: boolean = false;
+  private defaultWidth: string = 'auto';
+  private defaultHeight: string = '100%';
+  private defaultColorScheme: TreeViewColorTheme = TreeViewColorTheme.WHITE;
+  private nodeTreeHeight: number = 34;
+  // Default Throttle time fro scroll event
+  private defaultThrottleTime: number = 500;
+
+  constructor(
+    @Self() private el: ElementRef,
+    private cd: ChangeDetectorRef,
+    private treeViewUtils: FuiTreeViewUtilsService
+  ) {}
+
+  /**
+   * Initializes the Tree View component and all properties needed depending on inputs configuration
+   */
+  ngOnInit(): void {
+    this.serverSideComponent = this.dataRetriever.hasOwnProperty('getPagedChildNodeData');
+    this.treeViewStyles = {
+      width: this.config.width ? this.config.width : this.defaultWidth,
+      height: this.config.height ? this.config.height : this.defaultHeight,
+    };
+    this.scrollWidthChangeSub = this.treeViewUtils.treeViewScrollWidthChange.subscribe(() => this.updateScrollerWidth());
+    this.colorTheme = this.config.colorVariation ? this.config.colorVariation : this.defaultColorScheme;
+    this.border = this.hasBorders = this.config.hasBorders ? this.config.hasBorders : false;
+    if (this.serverSideComponent) {
+      this.bufferAmount = this.config.bufferAmount || this.bufferAmount;
+      const scrollThrottleTime = this.config.throttleTime || this.defaultThrottleTime;
+      this.scrollSubscription = this.vs.vsChange.pipe(throttleTime(scrollThrottleTime)).subscribe(pageInfo => {
```

Comment:
Gotcha, i took out the throttleTime now. 
I also added the cancelable promise we talked about. I added it as a promise array since on the server side we could potentially have a scrollPromise and a expandNode promise both going. The array takes care of any cancelable promises left